src/Sema.zig

```diff
@@ -17293,163 +17322,180 @@ fn beginComptimePtrMutation(
-                        .ty = elem_ty,
-                        .byte_offset = byte_offset,
-                        .is_mutable = parent.is_mutable,
-                    };
-                },
-                else => {
-                    if (elem_ptr.index != 0) {
-                        // TODO have the parent include the decl so we can say "declared here"
-                        return sema.fail(block, src, "out of bounds comptime load of index {d}", .{
-                            elem_ptr.index,
-                        });
+            const elem_ty = elem_ptr.elem_ty;
+
+            // An elem_ptr doesn't know the length or sentinel of its array_ptr, so we
+            // do an unchecked ptr load and then check the returned type ourselves
+            var deref = try beginComptimePtrLoad(sema, block, src, elem_ptr.array_ptr, null);
+
+            if (elem_ptr.index != 0) {
+                if (try sema.typeHasWellDefinedLayout(block, src, elem_ty)) {
+                    if (deref.parent) |*parent| {
+                        // Update the byte offset (in-place)
+                        const elem_size = try sema.typeAbiSize(block, src, elem_ty);
+                        const offset = parent.byte_offset + elem_size * elem_ptr.index;
+                        parent.byte_offset = try sema.usizeCast(block, src, offset);
                     }
-                    return ComptimePtrLoadKit{
-                        .root_val = parent.root_val,
-                        .root_ty = parent.root_ty,
-                        .val = parent.val,
-                        .ty = parent.ty,
-                        .byte_offset = parent.byte_offset,
-                        .is_mutable = parent.is_mutable,
-                    };
-                },
+                } else {
+                    deref.parent = null;
+                    deref.ty_without_well_defined_layout = elem_ty;
+                }
             }
+
+            // If we're loading an elem_ptr that was derived from a different type
+            // than the true type of the underlying decl, we cannot deref directly
+            const ty_matches = if (deref.target != null and deref.target.?.ty.isArrayLike()) x: {
+                const deref_elem_ty = deref.target.?.ty.childType();
+                break :x (try sema.coerceInMemoryAllowed(block, deref_elem_ty, elem_ty, false, target, src, src)) == .ok or
+                    (try sema.coerceInMemoryAllowed(block, elem_ty, deref_elem_ty, false, target, src, src)) == .ok;
+            } else false;
+            if (!ty_matches) {
+                deref.target = null;
+                break :blk deref;
+            }
+
+            var array_tv = deref.target.?;
+            const check_len = array_tv.ty.arrayLenIncludingSentinel();
+            if (elem_ptr.index >= check_len) {
+                // TODO have the deref include the decl so we can say "declared here"
```

Comment:
I'm uncertain about failing early here, rather than bitcasting. It is consistent with stage1 behavior, but it causes the following test to fail:

```zig
const S = extern struct {
    a: f32,
    b: [4] u8,
    c: f32,
};

comptime {
    const x = S{ 
        .a = 1.5,
        .b = [_]u8 { 1, 2, 3, 4 },
        .c = 2.6,
    };
    const ptr = &x.b[0];
    _ = @ptrCast([*c]const u8, ptr)[5]; // fails
}
```

It would be well-defined if we wanted to allow this load, but removing this error means we probably need to lift the length check to some callers of pointerDeref.

Reply:
Created #11157 for this one, since this test triggers a panic today.